File: evs360/contactor.cpp
#include <Arduino.h>

// Define constants
const int numPMs = 9; // Number of modules (PMs)
int assignments[numPMs]; // Array to track assignments of PMs to guns

void setup() {
    // Initialize GPIO pins for relays
    for (int pin = 2; pin < 2 + numPMs; pin++) {
        pinMode(pin, OUTPUT);
        digitalWrite(pin, LOW); // Ensure all relays are off initially
    }

    // Initialize assignments array
    for (int i = 0; i < numPMs; i++) {
        assignments[i] = -1; // No PM is assigned to any gun
    }
}

void activateRelays(int modules[], int moduleCount, int gunNumber) {
    int gunIdx = gunNumber - 1;

    // First, deactivate all relays connected to this gun
    for (int pmIdx = 0; pmIdx < numPMs; pmIdx++) {
        if (assignments[pmIdx] == gunIdx) {
            digitalWrite(2 + pmIdx, LOW); // Turn off the relay
            assignments[pmIdx] = -1; // Reset assignment
        }
    }

    // Then activate new connections
    for (int i = 0; i < moduleCount; i++) {
        int moduleNum = modules[i];
        int pmIdx = moduleNum - 1;

        // If this module is assigned to a different gun, deactivate that connection first
        if (assignments[pmIdx] != -1) {
            digitalWrite(2 + pmIdx, LOW); // Turn off the relay
            assignments[pmIdx] = -1; // Reset assignment
        }

        // Activate the new connection
        digitalWrite(2 + pmIdx, HIGH); // Turn on the relay
        assignments[pmIdx] = gunIdx; // Update assignment
    }
}

void loop() {
    // Example usage (replace with actual logic)
    int exampleModules[] = {1, 3, 5};
    activateRelays(exampleModules, 3, 1);
    delay(5000); // Wait for 5 seconds
}

File: evs360/contactor.py
def activate_relays(modules, gun_number, assignments, num_pms):
    gun_idx = gun_number - 1

    # First, deactivate all relays connected to this gun
    for pm_idx in range(num_pms):
        if assignments[pm_idx] == gun_idx:
            # Simulate deactivating the relay
            print(f"Deactivating relay for PM{pm_idx + 1} and Gun{gun_idx + 1}")
            assignments[pm_idx] = -1

    # Then activate new connections
    for module_num in modules:
        pm_idx = module_num - 1

        # If this module is assigned to a different gun, deactivate that connection first
        if assignments[pm_idx] != -1:
            print(f"Deactivating relay for PM{pm_idx + 1} from Gun{assignments[pm_idx] + 1}")
            assignments[pm_idx] = -1

        # Activate the new connection
        print(f"Activating relay for PM{pm_idx + 1} and Gun{gun_idx + 1}")
        assignments[pm_idx] = gun_idx

    # Update assignments display
    update_assignments(assignments)

def update_assignments(assignments):
    active_assignments = [
        f"PM{pm_idx + 1}→Gun{gun_idx + 1}" 
        for pm_idx, gun_idx in enumerate(assignments) 
        if gun_idx != -1
    ]
    if active_assignments:
        print("Active Assignments:", ", ".join(active_assignments))
    else:
        print("No active assignments")
